cleverhans/plot/image.py

```diff
@@ -0,0 +1,146 @@
+"""
+Functionality for displaying or saving images.
+"""
+from tempfile import mkstemp
+import os
+import platform
+
+import numpy as np
+from PIL import Image
+
+from cleverhans.utils import shell_call
+
+def show(ndarray, min_val=None, max_val=None):
+  """
+  Display an image.
+  :param ndarray: The image as an ndarray
+  :param min_val: The minimum pixel value in the image format
+  :param max_val: The maximum pixel valie in the image format
+    If min_val and max_val are not specified, attempts to
+    infer whether the image is in any of the common ranges:
+      [0, 1], [-1, 1], [0, 255]
+    This can be ambiguous, so it is better to specify if known.
+  """
+
+  # Create a temporary file with the suffix '.png'.
+  fd, path = mkstemp(suffix='.png')
+  os.close(fd)
+  save(path, ndarray, min_val, max_val)
+  shell_call(VIEWER_COMMAND + [path])
+
+
+
+def save(path, ndarray, min_val=None, max_val=None):
+  """
+  Save an image, represented as an ndarray, to the filesystem
+  :param path: string, filepath
+  :param ndarray: The image as an ndarray
+  :param min_val: The minimum pixel value in the image format
+  :param max_val: The maximum pixel valie in the image format
+    If min_val and max_val are not specified, attempts to
+    infer whether the image is in any of the common ranges:
+      [0, 1], [-1, 1], [0, 255]
+    This can be ambiguous, so it is better to specify if known.
+  """
+  as_pil(ndarray, min_val, max_val).save(path)
+
+def as_pil(ndarray, min_val=None, max_val=None):
+  """
+  Converts an ndarray to a PIL image.
+  :param ndarray: The numpy ndarray to convert
+  :param min_val: The minimum pixel value in the image format
+  :param max_val: The maximum pixel valie in the image format
+    If min_val and max_val are not specified, attempts to
+    infer whether the image is in any of the common ranges:
+      [0, 1], [-1, 1], [0, 255]
+    This can be ambiguous, so it is better to specify if known.
+  """
+
+  assert isinstance(ndarray, np.ndarray)
+
+  # rows x cols for grayscale image
+  # rows x cols x channels for color
+  assert ndarray.ndim in [2, 3]
+  if ndarray.ndim == 3:
+    channels = ndarray.shape[2]
+    # grayscale or RGB
+    assert channels in [1, 3]
+
+  actual_min = ndarray.min()
+  actual_max = ndarray.max()
+
+  if min_val is not None:
+    assert actual_min >= min_val
+    assert actual_max <= max_val
+
+  if 'float' in str(ndarray.dtype):
```

Comment:
It might be slightly better to use `np.issubdtype`

```suggestion
  if np.issubdtype(ndarray.dtype, np.floating):
```

Reply:
Thanks, that's a useful function. I didn't know about it.